### PR TITLE
Configuration edits

### PR DESCRIPTION
**Pull request by ryantheleach**
I feel like the configuration was slightly confusing, especially with how much we mentioned `default` when all it was getting was the recommended config locations rather then backing the configuration with inherited values. Still working my way through this, but making a PR in order to show progress and get feedback.

### CODE REVIEW

File: source/plugin/configuration/index.rst [hunk 1]
```diff
@@ -40,31 +40,34 @@ to store data, and they allow server administrators to customize plugin options
...
+returned pathname will be in a shared configuration directory. In that case, the configuration file for your plugin will
+be ``your_plugin_id.conf`` (with "your_plugin_id" replaced with your plugin's specified ID).
 
 .. tip::
 
     See :doc:`../main-class` for information on configuring your plugin ID.
 
-If you set ``sharedRoot`` to ``false``, the returned pathname will refer to a file named ``{pluginname}.conf`` in a
-directory specific to your plugin.
+If you set ``sharedRoot`` to ``false``, the returned pathname will refer to a file named ``pluginname.conf`` in a
```
Comment (saladoc) [inline source/plugin/configuration/index.rst]: your_plugin_id.conf

**Comment by ryantheleach** [comment]
from irc:

> 2:55 PM <Eliter> ok, on https://docs.spongepowered.org/en/plugin/configuration/loaders.html where do they suddenly get getConfigPath method?

**Comment by saladoc** [comment]
We need a better way of signalling "this variable is assumed to be present". should it just be `Path potentialFile = ...; // Assumed to be present`? That way it's clear to not be valid java so that may actually be clearer

**Comment by ryantheleach** [comment]
I think I can eliminate it completely, but yeah it would be good.

**Comment by ryantheleach** [comment]
Sorry all, I kinda just left this hanging in WIP state...

If someone could re-look over this with a fine toothed comb I'll finish up what I can just to get what's done so far merged in instead of waiting any longer.

File: source/plugin/configuration/nodes.rst [hunk 2]
```diff
@@ -183,3 +183,11 @@ with ``ConfigurationOptions`` that enable copying of default values and get an e
 line ``rootNode.getNode("modules", "blockCheats", "enabled").getBoolean(true)``. As the node does not yet exist,
 configurate creates it and writes the value ``true`` to it as per the ``ConfigurationOptions`` before returning it.
 When the config is then finished, the value ``true`` will persist on the node without ever being explicitly set.
+
+.. tip::
+
+    You may be tempted to use this to upgrade configs between versions or packaging a default configuration with your
+    plugin, however except for very basic configuration files this is likely to get awkward fast. at the end of this
```
Comment (Inscrutable) [inline source/plugin/configuration/nodes.rst]: I'd rephrase to ..."plugin, but this will quickly become awkward for all but the most basic configuration files."
Also, capitalise At

File: source/plugin/configuration/defaults.rst [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+    import java.net.URL;
+
+    URL jarConfigFile = this.getClass().getResource("defaultConfig.conf");
+    ConfigurationLoader<CommentedConfigurationNode> loader =
+      HoconConfigurationLoader.builder().setURL(jarConfigFile).build();
+
+For this example it is important to note that the ``getResource(...)`` method works relative to the location of the
+class it is called on. So if in the above example the class lies in the package ``me.username.myplugin``, the
+``defaultConfig.conf`` file must not lie in the jar file root, but instead in the directory ``me/username/myplugin``.
+This is usually accomplished by placing the ``defaultConfig.conf`` in a resources file in your project, be careful when
```
Comment (Inscrutable) [inline source/plugin/configuration/defaults.rst]: I suggest you break this into 2 sentences. Eg ..."your project. Be careful when"...
In general, please try not to let sentences get too long.

File: source/plugin/configuration/defaults.rst [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+    Since all Sponge plugins share a single namespace, all resources available to the ``getResource()`` method are
+    visible to all other plugins as well. Therefore, resources placed in the root of a jar may overwrite (or be
+    overwritten by) identically named resources in another jar. Placing those resources in unique folder structures
+    similar to your java packages will mitigate the danger of accidentally having a resource file overwritten by
+    another plugin.
+
+.. tip::
+
+    If you are having troubles getting your resource while in an IntelliJ dev environment, and you are not using 
+    ForgeGradle the resource files in your plugin may not be available on the classpath unless you enable the following
```
Comment (Inscrutable) [inline source/plugin/configuration/defaults.rst]: Long sentence needs a few more commas for clarity imho - after ForgeGradle, and options.

File: source/plugin/configuration/updating.rst [hunk 5]
```diff
@@ -0,0 +1,5 @@
+==========================
+Updating Versioned Configs
+==========================
+
+Words lots of words lots of words!
```
Comment (12AwsomeMan34) [inline source/plugin/configuration/updating.rst]: Any *update* on this?
(Ha ha, get the pun...?)

File: source/plugin/configuration/defaults.rst [hunk 6]
```diff
@@ -0,0 +1,49 @@
...
+.. figure:: /images/InheritProjectCompileOutputPath.png
+    :alt: a screenshot of IntelliJ Project Structure options.
+
+    File> Project Structure> Modules> ExamplePlugin> Paths> ``Inherit Project Compile Output Path``
+
+Now that you have your default configuration, we need to detect / merge it with the @DefaultConfig annotated
+configuration.
+
+
+
```
Comment (12AwsomeMan34) [inline source/plugin/configuration/defaults.rst]: Any specific reason for the extra lines?

File: source/plugin/configuration/defaults.rst [hunk 7]
```diff
@@ -0,0 +1,49 @@
...
+.. code-block:: groovy
+
+        idea { module { inheritOutputDirs = true } }
+
+.. figure:: /images/InheritProjectCompileOutputPath.png
+    :alt: a screenshot of IntelliJ Project Structure options.
+
+    File> Project Structure> Modules> ExamplePlugin> Paths> ``Inherit Project Compile Output Path``
+
+Now that you have your default configuration, we need to detect / merge it with the @DefaultConfig annotated
```
Comment (12AwsomeMan34) [inline source/plugin/configuration/defaults.rst]: Minor nitpick, but you could probably use the backticks \`\` around `@DefaultConfig`

File: source/plugin/configuration/defaults.rst [hunk 8]
```diff
@@ -0,0 +1,49 @@
+=========================
+Packaging Default Configs
+=========================
+
+Example: Loading a default config from the plugin jar file
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+.. code-block:: java
+
+    import java.net.URL;
```
Comment (12AwsomeMan34) [inline source/plugin/configuration/defaults.rst]: Any reason for excluding the imports for `ConfigurationLoader`, `CommentedConfigurationNode`, and `HoconConfigurationLoader`? This is the first code block for this page, or would they be deemed unnecessary?

Comment (RobertHerhold) [inline source/plugin/configuration/defaults.rst]: Yes, the imports are necessary.

**Comment by Minecrell** [comment]
@windy1 Can you check if any changes should be made here for the asset API?

File: source/plugin/configuration/defaults.rst [hunk 9]
```diff
@@ -0,0 +1,49 @@
...
+
+    URL jarConfigFile = this.getClass().getResource("defaultConfig.conf");
```
Comment (windy1) [inline source/plugin/configuration/defaults.rst]: This should be updated to use the Asset API.

**Comment by ryantheleach** [comment]
@eliterscripts Eliter> Can someone please edit the docs, to seriously improve the explaination of createEmptyNode in the configuration section of making a plugin?

**Comment by Inscrutable** [comment]
This is getting a bit ancient. Can we clear the dust off and get it finished, @ryantheleach ?

**Comment by Inscrutable** [comment]
All I want for Xmas is **THIS PR CLEANED UP AND MERGED.**
OK, that's a lie. *Please* @ryantheleach can this get resolved?

**Comment by Inscrutable** [comment]
One last call for @ryantheleach or someone else to rebase and finish this off.
Otherwise, I will close it on it's Anniversary.

**Comment by Tzky** [comment]
:(

**Comment by Inscrutable** [comment]
Oh well, thanks for trying. These notes could be revived if we have anyone else who actually understands this stuff. Until then, _they lie not dead but sleeping..._

**Comment by ryantheleach** [comment]
My PC is broken and wasn't going to be fixed before the deadline.

I'm sorry I just don't have the time committments currently, at least this shows the status of the ticket. I lost my train of thought on this ages ago, but kept meaning to essentially redo it.

Things that would have needed to be covered before I would have considered them finished would be,

Custom mappings, mapping annotations, upgrading config versions, loading configs from assets nicely and easily, recommending people make a backup of the previous config and not delete it before saving if an error occurs, Config Translators and how they fit into Data Translators, how to adapt from config translators to data translators without stepping on any other plugins toes by having multiple registrations.
